components/remerge/src/untyped_map.rs

```diff
@@ -0,0 +1,300 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! this module implements the untyped_map data type. This works as follows:
+//!
+//! - The "native" representation of an "untyped_map" value is just the
+//!   underlying map data, e.g. a `JsonObject`.
+//! - To convert a native map to any other format of map, the tombstone list
+//!   must be provided (of course, it's initially empty), creating an
+//!   `UntypedMap`
+//! - To convert a local map to a native map, the tombstone gets stored in the
+//!   database (if applicable) and then discarded.
+//!
+//! See the RFC for the merge algorithm for these.
+
+use crate::{error::*, JsonObject, JsonValue};
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, BTreeSet};
+
+pub type MapData = BTreeMap<String, JsonValue>;
+
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize, Default)]
+pub struct UntypedMap {
+    map: MapData,
+    #[serde(default)]
+    tombs: BTreeSet<String>,
+}
+
+// duplication is annoying here, but keeps the api clean and isn't that much
+// code
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum OnCollision {
+    /// Remove the map entry, keeping the tombstone (for prefer_deletion: true,
+    /// or if the tombstone)
+    DeleteEntry,
+    /// Keep the map entry, remove the tombstone (for prefer_deletion: false, or
+    /// if the data in the map is newer than tombstones, e.g. when updating a
+    /// record with new data).
+    KeepEntry,
+}
+
+impl From<OnCollision> for CollisionHandling {
+    fn from(src: OnCollision) -> Self {
+        match src {
+            OnCollision::DeleteEntry => Self::DeleteEntry,
+            OnCollision::KeepEntry => Self::KeepEntry,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+enum CollisionHandling {
+    /// Emit a `UntypedMapTombstoneCollision` error (a decent default if the map
+    /// data and tombstones shouldn't collide).
+    Error,
+    /// See OnCollision::DeleteEntry
+    DeleteEntry,
+    /// See OnCollision::KeepEntry
+    KeepEntry,
+}
+
+impl UntypedMap {
+    pub fn empty() -> Self {
+        Self::default()
+    }
+
+    pub fn from_native(map: JsonObject) -> Self {
+        Self {
+            map: map.into_iter().collect(),
+            tombs: Default::default(),
+        }
+    }
+
+    fn new_impl(
+        mut map: MapData,
+        mut tombs: BTreeSet<String>,
+        on_tombstone_map_collision: CollisionHandling,
+    ) -> Result<UntypedMap> {
+        // Should usually be empty so the cloning probably isn't worth fighting
+        let collided = map
+            .keys()
+            .filter(|k| tombs.contains(k.as_str()))
+            .cloned()
+            .collect::<Vec<_>>();
```

Comment:
Why do we need to `collect`?

Reply:
Actually we do, since otherwise we have an immutable and mutbale borrow out at the same time.